### ISSUE TITLE
state_map_iset__: illegal state transition

### ISSUE THREAD

**Issue by sondreso** [equinor/ert#3113]
**Describe the bug**
User reported: 

> Hi,
>
> I have a model that I'm attempting to run, but I'm running into the same error past a certain number of realizations (somewhere between 1000-2000). The error makes me me think that I'm overflowing an int somewhere, but I'm not sure exactly how the error message is calculating realizations (I don't have anywhere near 32523 in the example below). 
>
> Could anyone offer some insight on this issue? It seems like I'm running into a limit, but I'm not sure if I can restructure something or if I can at least assess what limit Im'm hitting.
>
> Error message: state_map_iset__: illegal state transition for realisation:32523 1 -> 4 
>
> Thanks!

Here is the stack trace: 

```
2022-03-11 12:05:56,939 - ert_logging._log_util_abort - ERROR - C trace:
Stack trace (most recent call last) in thread 3734:
#31   Object "/lib64/libpython3.6m.so.1.0", at 0x7f0c2d2ddb52, in PyEval_EvalFrameEx
#30   Object "/lib64/libpython3.6m.so.1.0", at 0x7f0c2d2dd739, in PyEval_EvalFrameEx
#29   Object "/lib64/libpython3.6m.so.1.0", at 0x7f0c2d2d3dec, in _PyEval_EvalFrameDefault
#28   Object "/lib64/libpython3.6m.so.1.0", at 0x7f0c2d233c02, in PyObject_Call
#27   Object "/lib64/libpython3.6m.so.1.0", at 0x7f0c2d233f90, in _PyObject_Call_Prepend
#26   Object "/lib64/libpython3.6m.so.1.0", at 0x7f0c2d233e7d, in _PyObject_FastCallDict
#25   Object "/lib64/libpython3.6m.so.1.0", at 0x7f0c2d2dec41, in _PyFunction_FastCallDict
#24   Object "/lib64/libpython3.6m.so.1.0", at 0x7f0c2d2d2146, in _PyEval_EvalFrameDefault
#23   Object "/lib64/libpython3.6m.so.1.0", at 0x7f0c2d2ddb52, in PyEval_EvalFrameEx
#22   Object "/lib64/libpython3.6m.so.1.0", at 0x7f0c2d2dd739, in PyEval_EvalFrameEx
#21   Object "/lib64/libpython3.6m.so.1.0", at 0x7f0c2d2d2146, in _PyEval_EvalFrameDefault
#20   Object "/lib64/libpython3.6m.so.1.0", at 0x7f0c2d2ddb52, in PyEval_EvalFrameEx
#19   Object "/lib64/libpython3.6m.so.1.0", at 0x7f0c2d2dd739, in PyEval_EvalFrameEx
#18   Object "/lib64/libpython3.6m.so.1.0", at 0x7f0c2d2d2146, in _PyEval_EvalFrameDefault
#17   Object "/lib64/libpython3.6m.so.1.0", at 0x7f0c2d2ddbdb, in PyEval_EvalFrameEx
#16   Object "/lib64/libpython3.6m.so.1.0", at 0x7f0c2d233dff, in _PyObject_FastCallDict
#15   Object "/lib64/libpython3.6m.so.1.0", at 0x7f0c2d287b9c, in _PyObject_LookupSpecial
#14   Object "/lib64/libpython3.6m.so.1.0", at 0x7f0c2d233c02, in PyObject_Call
#13   Object "/lib64/libpython3.6m.so.1.0", at 0x7f0c2d233f90, in _PyObject_Call_Prepend
#12   Object "/lib64/libpython3.6m.so.1.0", at 0x7f0c2d233e7d, in _PyObject_FastCallDict
#11   Object "/lib64/libpython3.6m.so.1.0", at 0x7f0c2d2ded2e, in _PyFunction_FastCallDict
#10   Object "/lib64/libpython3.6m.so.1.0", at 0x7f0c2d2dc9a3, in PyEval_EvalFrameEx
#9    Object "/lib64/libpython3.6m.so.1.0", at 0x7f0c2d2d3dec, in _PyEval_EvalFrameDefault
#8    Object "/lib64/libpython3.6m.so.1.0", at 0x7f0c2d233c02, in PyObject_Call
#7    Object "/usr/lib64/python3.6/lib-dynload/_ctypes.cpython-36m-x86_64-linux-gnu.so", at 0x7f0bc41887c5, in _init
#6    Object "/usr/lib64/python3.6/lib-dynload/_ctypes.cpython-36m-x86_64-linux-gnu.so", at 0x7f0bc418f67f, in _ctypes_callproc
#5    Object "/lib64/libffi.so.6", at 0x7f0bc3f7c714, in ffi_call
#4    Object "/lib64/libffi.so.6", at 0x7f0bc3f7cdeb, in ffi_call_unix64
#3    Object "/prog/res/komodo/2022.02.04-py36-rhel7/root/lib64/python3.6/site-packages/res/_lib.cpython-36m-x86_64-linux-gnu.so", at 0x7f0c2223ac27, in enkf_state_complete_forward_modelOK
#2    Object "/prog/res/komodo/2022.02.04-py36-rhel7/root/lib64/python3.6/site-packages/res/_lib.cpython-36m-x86_64-linux-gnu.so", at 0x7f0c2223a981, in ensemble_config_get_or_create_summary_node
#1    Object "/prog/res/komodo/2022.02.04-py36-rhel7/root/lib64/python3.6/site-packages/res/_lib.cpython-36m-x86_64-linux-gnu.so", at 0x7f0c2225488a, in state_map_iset
#0    Object "/prog/res/komodo/2022.02.04-py36-rhel7/root/lib64/python3.6/site-packages/ecl/.libs/libecl.so", at 0x7f0c22924cae, in util_abort__
```

This also happened with index 32737 and 32626 as well. 

The error originates from this function: 

https://github.com/equinor/ert/blob/e4ec2b1b1534ce89c4703c718ab3e94a3221ca33/libres/lib/enkf/state_map.cpp#L145-L156

The function `int_vector_safe_iget` returns a default value (1=STATE_UNDEFINED) if the index is out of bounds, so the problem is most likely that the index is to large, and not that the state transition is incorrect. We should try to figure out where this is coming from. 

**To Reproduce**

Unknown

**Expected behavior**
No crash

**Enviromment**
 - OS: RHEL7
 - ERT/Komodo Release: 2022.02.04
 - Python version: 3.6
 - Remote/HPC execution involved: [yes|no]

**Additional context**
Link to logs can be obtained by contacting @sondreso

**Comment by jondequinor** [comment]
I'm going to close this as a duplicate of https://github.com/equinor/ert/issues/2781 since that's the most relevant cause for this issue.

**Comment by BjarneHerland** [comment]
I'm reopening this for now because we see several issues most likely caused by `run_arg` being collected, as commented above by @jondequinor

**Comment by oyvindeide** [comment]
Note that in the new case we get: 
```
ert_logging._log_util_abort._log_util_abort.<locals>.UtilAbort: state_map_iset__: illegal state transition for realisation:32513 1 -> 8
```
which is a subtle difference, it is trying to move to `STATE_LOAD_FAILURE = 8`, because the loading of data failed. Might not be relevant.

**Comment by BjarneHerland** [comment]
So, here is what I think happens:

`RunArg`-objects are not instantiated from Python but always retrieved by indexing an `ErtRunContext`-object

https://github.com/equinor/ert/blob/27fe7774b9e6a846c9798fb640694291b3e98a8c/res/enkf/ert_run_context.py#L202-L212

Furthermore, it looks like a `ErtRunContext`-object can be GCd even if Python holds references to individual `RunArg`-objects retrieved from it. I'll assume this for now - @jondequinor , @sondreso or @oyvindeide : you might be able to verify/falsify this assumption?

When preparing an ensemble, a `ErtRunContext`-object eventually ends up in a loop retrieving a `RunArg` python-object for each realization

https://github.com/equinor/ert/blob/27fe7774b9e6a846c9798fb640694291b3e98a8c/ert/ensemble_evaluator/builder/_ensemble_builder.py#L120

storing these as parameters to callback-functions

https://github.com/equinor/ert/blob/27fe7774b9e6a846c9798fb640694291b3e98a8c/ert/ensemble_evaluator/builder/_ensemble_builder.py#L136-L140

One example from a logfile looks like this

`
2022-04-09 02:07:27,676 - res.enkf - INFO - [32730:0000-0000] Forward model complete - starting to load results.
`

which comes from line 602 in the "done" callback-function with the `RunArg`-object as parameter

https://github.com/equinor/ert/blob/27fe7774b9e6a846c9798fb640694291b3e98a8c/libres/lib/enkf/enkf_state.cpp#L586-L604

`iens` in line 593 most likely comes from a freed memory-area (see also the examples from @jondequinor above).

Exactly **why** this can happen is not entirely clear to me: I would expect all callbacks for all realizations to have finished before starting a new iteration, but I think this points in the direction of the problem.

**Comment by BjarneHerland** [comment]
The number of log-entries containing these large `iens`-values have decreased dramatically from 315 entries in April to 9 so far in May, found by counting log-entries in the Azure dashboard using this expression

> let Failures = AppTraces | where indexof_regex(Properties, @"\[32(\d){2,5}:", 0) > 1;

There is also a chance that such issues are caused by #3401 crashing parts of ERT while leaving other parts running, although this is a speculation.

Since this issue is hard to reproduce I suggest to put it in the backlog, monitor occurrences via Azure, and pick it up again if it turns out to be a problem.

**Comment by oyvindeide** [comment]
We have not had any more reports on this, and the whole ownership of `RunArg` though `RunContext` has been changed, so I will close this.
